openedx/core/djangoapps/user_api/models.py

```
@@ -25,28 +56,79 @@ class Meta:  # pylint: disable=missing-docstring
-        user_pref.value = preference_value
-        user_pref.save()
+        user_preference, _ = cls.objects.get_or_create(user=user, key=preference_key)
+        old_value = user_preference.value
+        user_preference.value = preference_value
+        try:
+            user_preference.full_clean()
+        finally:
+            user_preference.value = old_value
 
     @classmethod
     def get_preference(cls, user, preference_key, default=None):
-        """
-        Gets the user preference value for a given key
+        """Gets the user preference value for a given key
 
-        Returns the given default if there isn't a preference for the given key
-        """
+        Arguments:
+            user (User): The user whose preference should be set.
+            preference_key (string): The key for the user preference.
+            default (object): The default value to return if the preference is not set.
 
+        Returns:
+            The user preference value, or the specified default if one is not set.
+        """
         try:
-            user_pref = cls.objects.get(user=user, key=preference_key)
-            return user_pref.value
+            user_preference = cls.objects.get(user=user, key=preference_key)
+            return user_preference.value
         except cls.DoesNotExist:
             return default
 
+    @classmethod
+    def delete_preference(cls, user, preference_key):
+        """Deletes the user preference value for a given key
+
+        Arguments:
+            user (User): The user whose preference should be set.
+            preference_key (string): The key for the user preference.
+
+        Raises:
+            PreferenceNotFound: No preference was found with the given key.
+        """
+        try:
+            user_preference = cls.objects.get(user=user, key=preference_key)
+        except cls.DoesNotExist:
+            raise PreferenceNotFound()
```

Again, I'm not sure we should raise an error in this case.